query/marshal.go

```diff
@@ -0,0 +1,189 @@
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"unsafe"
+)
+
+// We implement a custom binary marshaller for a list of repos to
+// branches. When profiling Sourcegraph this is one of the dominant items.
+//
+// Wire-format of map[string][]string is pretty straightforward:
+//
+// byte(1) version
+// uvarint(len(map))
+// for k, vs in map:
+//   str(k)
+//   uvarint(len(vs))
+//   for v in vs:
+//     str(v)
+//
+//  where str(v) is uvarint(len(v)) bytes(v)
+//
+// The above format gives about the same size encoding as gob does. However,
+// gob doesn't have a specialization for map[string][]string so we get to
+// avoid a lot of intermediate allocations.
+//
+// The only other specialization we add is treating []string{"HEAD"} as if it
+// was []string{}. This is the most common value for branches so avoids the
+// need to write it on the wire. This makes us beat gob for encoded size.
+//
+// The above adds up to a huge improvement, worth the extra complexity:
+//
+// name                   old time/op    new time/op    delta
+// RepoBranches_Encode-8    2.37ms ± 3%    0.66ms ± 2%  -72.02%  (p=0.000 n=10+10)
+// RepoBranches_Decode-8    4.19ms ± 2%    0.75ms ± 2%  -82.17%  (p=0.000 n=10+10)
+//
+// name                   old bytes      new bytes      delta
+// RepoBranches_Encode-8     393kB ± 0%     344kB ± 0%  -12.48%  (p=0.000 n=10+10)
+//
+// name                   old alloc/op   new alloc/op   delta
+// RepoBranches_Encode-8     726kB ± 0%     344kB ± 0%  -52.59%  (p=0.000 n=10+8)
+// RepoBranches_Decode-8    2.31MB ± 0%    1.44MB ± 0%  -37.51%  (p=0.000 n=9+10)
+//
+// name                   old allocs/op  new allocs/op  delta
+// RepoBranches_Encode-8     20.0k ± 0%      0.0k ± 0%  -99.99%  (p=0.000 n=10+10)
+// RepoBranches_Decode-8     50.6k ± 0%      0.4k ± 0%  -99.26%  (p=0.000 n=10+10)
+
+// repoBranchesEncode implements an efficient encoder for RepoBranches.
+func repoBranchesEncode(repoBranches map[string][]string) ([]byte, error) {
+	var b bytes.Buffer
+	var enc [8]byte
+	varint := func(n int) {
+		m := binary.PutUvarint(enc[:], uint64(n))
+		b.Write(enc[:m])
+	}
+	str := func(s string) {
+		varint(len(s))
+		b.WriteString(s)
+	}
+	strSize := func(s string) int {
+		return binary.PutUvarint(enc[:], uint64(len(s))) + len(s)
+	}
+
+	// Version
+	b.WriteByte(1)
+
+	// Length
+	varint(len(repoBranches))
+
+	// Calculate size
+	size := 0
+	for name, branches := range repoBranches {
+		size += strSize(name) + 1
+		if l := len(branches); l == 1 && branches[0] == "HEAD" {
+			continue
+		} else if l > 255 {
+			return nil, fmt.Errorf("can't encode more than 255 branches: %d", l)
+		}
+		for _, branch := range branches {
+			size += strSize(branch)
+		}
+	}
+	b.Grow(size)
+
+	for name, branches := range repoBranches {
+		str(name)
+
+		// Special case "HEAD"
+		if len(branches) == 1 && branches[0] == "HEAD" {
+			b.WriteByte(0)
+			continue
+		}
+
+		// length of branches is 64 or less
+		b.WriteByte(byte(len(branches)))
```

Comment:
I could just use varint here like elsewhere. varint encodes numbers less than 127 as a single byte anyways. then we don't need to worry about the 255 issue. I think I will update the code to do this (and benchmark)

Reply:
using varint for this length increases the runtime by 12%! I'll keep it as is :)